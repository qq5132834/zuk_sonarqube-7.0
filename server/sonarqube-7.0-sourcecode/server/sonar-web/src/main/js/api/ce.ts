/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { getJSON, post, RequestData } from '../helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';

export interface PendingTask {
  componentKey: string;
  componentName: string;
  componentQualifier: string;
  id: string;
  logs: boolean;
  organization: string;
  status: string;
  submittedAt: Date;
  submitterLogin?: string;
  type: string;
}

export interface Task extends PendingTask {
  analysisId?: string;
  errorMessage?: string;
  errorType?: string;
  executionTimeMs: number;
  executedAt: Date;
  hasErrorStacktrace: boolean;
  hasScannerContext: boolean;
  startedAt: Date;
}

export function getActivity(data: RequestData): Promise<any> {
  return getJSON('/api/ce/activity', data);
}

export function getStatus(componentId?: string): Promise<any> {
  const data = {};
  if (componentId) {
    Object.assign(data, { componentId });
  }
  return getJSON('/api/ce/activity_status', data);
}

export function getTask(id: string, additionalFields?: string[]): Promise<any> {
  return getJSON('/api/ce/task', { id, additionalFields }).then(r => r.task);
}

export function cancelTask(id: string): Promise<any> {
  return post('/api/ce/cancel', { id }).then(() => getTask(id), () => getTask(id));
}

export function cancelAllTasks(): Promise<any> {
  return post('/api/ce/cancel_all');
}

export function getTasksForComponent(
  componentKey: string
): Promise<{ queue: PendingTask[]; current: Task }> {
  return getJSON('/api/ce/component', { componentKey }).catch(throwGlobalError);
}

export function getTypes(): Promise<any> {
  return getJSON('/api/ce/task_types').then(r => r.taskTypes);
}

export function getWorkers(): Promise<{ canSetWorkerCount: boolean; value: number }> {
  return getJSON('/api/ce/worker_count').catch(throwGlobalError);
}

export function setWorkerCount(count: number): Promise<void | Response> {
  return post('/api/ce/set_worker_count', { count }).catch(throwGlobalError);
}
