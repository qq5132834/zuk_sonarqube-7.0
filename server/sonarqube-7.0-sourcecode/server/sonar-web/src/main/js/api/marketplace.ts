/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { checkStatus, corsRequest, getJSON, parseJSON, post, postJSON } from '../helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';

export interface Edition {
  key: string;
  name: string;
  textDescription: string;
  homeUrl: string;
  licenseRequestUrl: string;
  downloadUrl: string;
}

export interface EditionsPerVersion {
  [version: string]: Edition[];
}

export interface EditionStatus {
  currentEditionKey?: string;
  nextEditionKey?: string;
  installError?: string;
  installationStatus:
    | 'NONE'
    | 'AUTOMATIC_IN_PROGRESS'
    | 'MANUAL_IN_PROGRESS'
    | 'AUTOMATIC_READY'
    | 'UNINSTALL_IN_PROGRESS';
}

export function getEditionStatus(): Promise<EditionStatus> {
  return getJSON('/api/editions/status');
}

export function getEditionsList(url: string): Promise<EditionsPerVersion> {
  return corsRequest(url)
    .submit()
    .then(checkStatus)
    .then(parseJSON);
}

export function getLicensePreview(data: {
  license: string;
}): Promise<{
  nextEditionKey: string;
  previewStatus: 'NO_INSTALL' | 'AUTOMATIC_INSTALL' | 'MANUAL_INSTALL';
}> {
  return postJSON('/api/editions/preview', data).catch(throwGlobalError);
}

export function getFormData(): Promise<{ serverId: string; ncloc: number }> {
  return getJSON('/api/editions/form_data').catch(throwGlobalError);
}

export function applyLicense(data: { license: string }): Promise<EditionStatus> {
  return postJSON('/api/editions/apply_license', data).catch(throwGlobalError);
}

export function uninstallEdition(): Promise<void | Response> {
  return post('/api/editions/uninstall').catch(throwGlobalError);
}

export function dismissErrorMessage(): Promise<void | Response> {
  return post('/api/editions/clear_error_message').catch(throwGlobalError);
}
