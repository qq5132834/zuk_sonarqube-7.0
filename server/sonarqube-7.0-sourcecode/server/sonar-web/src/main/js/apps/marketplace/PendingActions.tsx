/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import RestartForm from '../../components/common/RestartForm';
import { cancelPendingPlugins, PluginPending } from '../../api/plugins';
import { translate } from '../../helpers/l10n';

interface Props {
  pending: {
    installing: PluginPending[];
    updating: PluginPending[];
    removing: PluginPending[];
  };
  refreshPending: () => void;
}

interface State {
  openRestart: boolean;
}

export default class PendingActions extends React.PureComponent<Props, State> {
  state: State = { openRestart: false };

  handleOpenRestart = () => this.setState({ openRestart: true });
  hanleCloseRestart = () => this.setState({ openRestart: false });

  handleRevert = () => {
    cancelPendingPlugins().then(this.props.refreshPending, () => {});
  };

  render() {
    const { installing, updating, removing } = this.props.pending;
    const hasPendingActions = installing.length || updating.length || removing.length;
    if (!hasPendingActions) {
      return null;
    }

    return (
      <div className="js-pending alert alert-warning">
        <div className="display-inline-block">
          <p>{translate('marketplace.sonarqube_needs_to_be_restarted_to')}</p>
          <ul className="list-styled spacer-top">
            {installing.length > 0 && (
              <li>
                <FormattedMessage
                  defaultMessage={translate('marketplace.install_x_plugins')}
                  id="marketplace.install_x_plugins"
                  values={{ nb: <strong>{installing.length}</strong> }}
                />
              </li>
            )}
            {updating.length > 0 && (
              <li>
                <FormattedMessage
                  defaultMessage={translate('marketplace.update_x_plugins')}
                  id="marketplace.update_x_plugins"
                  values={{ nb: <strong>{updating.length}</strong> }}
                />
              </li>
            )}
            {removing.length > 0 && (
              <li>
                <FormattedMessage
                  defaultMessage={translate('marketplace.uninstall_x_plugins')}
                  id="marketplace.uninstall_x_plugins"
                  values={{ nb: <strong>{removing.length}</strong> }}
                />
              </li>
            )}
          </ul>
        </div>
        <div className="pull-right">
          <button className="js-restart little-spacer-right" onClick={this.handleOpenRestart}>
            {translate('marketplace.restart')}
          </button>
          <button className="js-cancel-all button-red" onClick={this.handleRevert}>
            {translate('marketplace.revert')}
          </button>
        </div>
        {this.state.openRestart && <RestartForm onClose={this.hanleCloseRestart} />}
      </div>
    );
  }
}
