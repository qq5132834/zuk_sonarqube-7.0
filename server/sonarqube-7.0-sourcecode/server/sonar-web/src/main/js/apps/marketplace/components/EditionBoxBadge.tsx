/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import CheckIcon from '../../../components/icons-components/CheckIcon';
import { EditionStatus } from '../../../api/marketplace';
import { translate } from '../../../helpers/l10n';

interface Props {
  editionKey: string;
  status: EditionStatus;
}

export default function EditionBoxBadge({ editionKey, status }: Props) {
  const isInstalled =
    status.currentEditionKey === editionKey ||
    (!status.currentEditionKey && editionKey === 'community');
  const isProgressing =
    status.nextEditionKey === editionKey || (!status.nextEditionKey && editionKey === 'community');
  const inProgressStatus = [
    'AUTOMATIC_READY',
    'AUTOMATIC_IN_PROGRESS',
    'UNINSTALL_IN_PROGRESS'
  ].includes(status.installationStatus);

  if (inProgressStatus) {
    if (isProgressing) {
      return (
        <span className="marketplace-edition-badge badge badge-normal-size">
          {status.installationStatus === 'AUTOMATIC_IN_PROGRESS'
            ? translate('marketplace.installing')
            : translate('marketplace.pending')}
        </span>
      );
    }
  } else if (isInstalled) {
    return (
      <span className="marketplace-edition-badge badge badge-normal-size display-flex-center">
        <CheckIcon size={14} className="little-spacer-right" />
        {translate('marketplace.installed')}
      </span>
    );
  }

  return null;
}
