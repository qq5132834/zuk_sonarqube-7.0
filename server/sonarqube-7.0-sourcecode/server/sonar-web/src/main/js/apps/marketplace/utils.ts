/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { memoize, sortBy } from 'lodash';
import { Plugin, PluginAvailable, PluginInstalled, PluginPending } from '../../api/plugins';
import { cleanQuery, parseAsString, RawQuery, serializeString } from '../../helpers/query';
import { Edition } from '../../api/marketplace';

export interface Query {
  filter: string;
  search?: string;
}

export function isPluginAvailable(plugin: Plugin): plugin is PluginAvailable {
  return (plugin as any).release !== undefined;
}

export function isPluginInstalled(plugin: Plugin): plugin is PluginInstalled {
  return isPluginPending(plugin) && (plugin as any).updatedAt !== undefined;
}

export function isPluginPending(plugin: Plugin): plugin is PluginPending {
  return (plugin as any).version !== undefined;
}

export function filterPlugins(plugins: Plugin[], search: string): Plugin[] {
  const s = search.toLowerCase();
  return plugins.filter(plugin => {
    return (
      plugin.name.toLowerCase().includes(s) ||
      plugin.description.toLowerCase().includes(s) ||
      (plugin.category || '').toLowerCase().includes(s)
    );
  });
}

const EDITIONS_ORDER = ['community', 'developer', 'enterprise', 'datacenter'];
export function sortEditions(editions: Edition[]): Edition[] {
  return sortBy(editions, edition => EDITIONS_ORDER.indexOf(edition.key));
}

export const DEFAULT_FILTER = 'all';
export const parseQuery = memoize((urlQuery: RawQuery): Query => ({
  filter: parseAsString(urlQuery['filter']) || DEFAULT_FILTER,
  search: parseAsString(urlQuery['search'])
}));

export const serializeQuery = memoize((query: Query): RawQuery =>
  cleanQuery({
    filter: query.filter === DEFAULT_FILTER ? undefined : serializeString(query.filter),
    search: query.search ? serializeString(query.search) : undefined
  })
);
