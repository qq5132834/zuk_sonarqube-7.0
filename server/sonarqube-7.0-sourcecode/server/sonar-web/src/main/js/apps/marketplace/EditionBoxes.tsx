/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import EditionBox from './components/EditionBox';
import LicenseEditionForm from './components/LicenseEditionForm';
import UninstallEditionForm from './components/UninstallEditionForm';
import { Edition, EditionStatus } from '../../api/marketplace';
import { translate } from '../../helpers/l10n';
import { sortEditions } from './utils';

export interface Props {
  canInstall: boolean;
  canUninstall: boolean;
  editions?: Edition[];
  editionStatus?: EditionStatus;
  loading: boolean;
  updateCenterActive: boolean;
  updateEditionStatus: (editionStatus: EditionStatus) => void;
}

interface State {
  installEdition?: Edition;
  openUninstallForm: boolean;
}

export default class EditionBoxes extends React.PureComponent<Props, State> {
  state: State = { openUninstallForm: false };

  handleOpenLicenseForm = (edition: Edition) => this.setState({ installEdition: edition });
  handleCloseLicenseForm = () => this.setState({ installEdition: undefined });

  handleOpenUninstallForm = () => this.setState({ openUninstallForm: true });
  handleCloseUninstallForm = () => this.setState({ openUninstallForm: false });

  renderForms(sortedEditions: Edition[], installedIdx?: number) {
    const { canInstall, canUninstall, editionStatus } = this.props;
    const { installEdition, openUninstallForm } = this.state;
    const installEditionIdx =
      installEdition && sortedEditions.findIndex(edition => edition.key === installEdition.key);

    if (canInstall && installEdition) {
      return (
        <LicenseEditionForm
          edition={installEdition}
          editions={sortedEditions}
          isDowngrade={
            installedIdx !== undefined &&
            installEditionIdx !== undefined &&
            installEditionIdx < installedIdx
          }
          onClose={this.handleCloseLicenseForm}
          updateEditionStatus={this.props.updateEditionStatus}
        />
      );
    }

    if (canUninstall && openUninstallForm && editionStatus && editionStatus.currentEditionKey) {
      return (
        <UninstallEditionForm
          edition={sortedEditions.find(edition => edition.key === editionStatus.currentEditionKey)}
          editionStatus={editionStatus}
          onClose={this.handleCloseUninstallForm}
          updateEditionStatus={this.props.updateEditionStatus}
        />
      );
    }

    return null;
  }

  render() {
    const { canInstall, canUninstall, editions, loading } = this.props;

    if (loading) {
      return <i className="big-spacer-bottom spinner" />;
    }

    if (!editions) {
      return (
        <div className="spacer-bottom marketplace-editions">
          <span className="alert alert-info">
            <FormattedMessage
              defaultMessage={translate('marketplace.editions_unavailable')}
              id="marketplace.editions_unavailable"
              values={{
                url: (
                  <a href="https://redirect.sonarsource.com/editions/editions.html" target="_blank">
                    SonarSource.com
                  </a>
                )
              }}
            />
          </span>
        </div>
      );
    }

    const sortedEditions = sortEditions(editions);
    const status = this.props.editionStatus || { installationStatus: 'NONE' };
    const inProgressStatus = [
      'AUTOMATIC_IN_PROGRESS',
      'AUTOMATIC_READY',
      'UNINSTALL_IN_PROGRESS'
    ].includes(status.installationStatus);
    const installedIdx = sortedEditions.findIndex(
      edition => edition.key === status.currentEditionKey
    );
    const nextIdx = sortedEditions.findIndex(edition => edition.key === status.nextEditionKey);
    const currentIdx = inProgressStatus ? nextIdx : installedIdx;
    return (
      <div className="spacer-bottom marketplace-editions">
        <EditionBox
          actionLabel={translate('marketplace.downgrade')}
          disableAction={inProgressStatus}
          displayAction={canUninstall && currentIdx > 0}
          edition={sortedEditions[0]}
          editionStatus={status}
          key={sortedEditions[0].key}
          onAction={this.handleOpenUninstallForm}
        />
        {sortedEditions
          .slice(1)
          .map((edition, idx) => (
            <EditionBox
              actionLabel={
                currentIdx > idx + 1
                  ? translate('marketplace.downgrade')
                  : translate('marketplace.upgrade')
              }
              disableAction={inProgressStatus}
              displayAction={canInstall && currentIdx !== idx + 1}
              edition={edition}
              editionStatus={status}
              key={edition.key}
              onAction={this.handleOpenLicenseForm}
            />
          ))}

        {this.renderForms(sortedEditions, installedIdx)}
      </div>
    );
  }
}
