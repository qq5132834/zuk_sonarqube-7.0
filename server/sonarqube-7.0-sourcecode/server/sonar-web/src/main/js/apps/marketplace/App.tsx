/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as PropTypes from 'prop-types';
import { sortBy, uniqBy } from 'lodash';
import Helmet from 'react-helmet';
import Header from './Header';
import EditionBoxes from './EditionBoxes';
import Footer from './Footer';
import PendingActions from './PendingActions';
import PluginsList from './PluginsList';
import Search from './Search';
import {
  getAvailablePlugins,
  getInstalledPluginsWithUpdates,
  getPendingPlugins,
  getPluginUpdates,
  Plugin,
  PluginPending
} from '../../api/plugins';
import { Edition, EditionStatus } from '../../api/marketplace';
import { RawQuery } from '../../helpers/query';
import { translate } from '../../helpers/l10n';
import { filterPlugins, parseQuery, Query, serializeQuery } from './utils';
import './style.css';

export interface Props {
  editions?: Edition[];
  editionsReadOnly: boolean;
  editionStatus?: EditionStatus;
  loadingEditions: boolean;
  location: { pathname: string; query: RawQuery };
  standaloneMode: boolean;
  updateCenterActive: boolean;
  setEditionStatus: (editionStatus: EditionStatus) => void;
}

interface State {
  loadingPlugins: boolean;
  pending: {
    installing: PluginPending[];
    updating: PluginPending[];
    removing: PluginPending[];
  };
  plugins: Plugin[];
}

export default class App extends React.PureComponent<Props, State> {
  mounted: boolean;

  static contextTypes = {
    router: PropTypes.object.isRequired
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      loadingPlugins: true,
      pending: {
        installing: [],
        updating: [],
        removing: []
      },
      plugins: []
    };
  }

  componentDidMount() {
    this.mounted = true;
    this.fetchPendingPlugins();
    this.fetchQueryPlugins();
  }

  componentDidUpdate(prevProps: Props) {
    if (prevProps.location.query.filter !== this.props.location.query.filter) {
      this.fetchQueryPlugins();
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  fetchQueryPlugins = () => {
    const query = parseQuery(this.props.location.query);
    if (query.filter === 'updates') {
      this.fetchUpdatesOnly();
    } else {
      this.fetchAllPlugins();
    }
  };

  fetchAllPlugins = () => {
    this.setState({ loadingPlugins: true });
    Promise.all([getInstalledPluginsWithUpdates(), getAvailablePlugins()]).then(
      ([installed, available]) => {
        if (this.mounted) {
          this.setState({
            loadingPlugins: false,
            plugins: sortBy(uniqBy([...installed, ...available.plugins], 'key'), 'name')
          });
        }
      },
      this.stopLoadingPlugins
    );
  };

  fetchUpdatesOnly = () => {
    this.setState({ loadingPlugins: true });
    getPluginUpdates().then(plugins => {
      if (this.mounted) {
        this.setState({ loadingPlugins: false, plugins });
      }
    }, this.stopLoadingPlugins);
  };

  fetchPendingPlugins = () =>
    getPendingPlugins().then(
      pending => {
        if (this.mounted) {
          this.setState({ pending });
        }
      },
      () => {}
    );

  updateQuery = (newQuery: Partial<Query>) => {
    const query = serializeQuery({ ...parseQuery(this.props.location.query), ...newQuery });
    this.context.router.push({ pathname: this.props.location.pathname, query });
  };

  stopLoadingPlugins = () => {
    if (this.mounted) {
      this.setState({ loadingPlugins: false });
    }
  };

  render() {
    const { editions, editionStatus, standaloneMode } = this.props;
    const { loadingPlugins, plugins, pending } = this.state;
    const query = parseQuery(this.props.location.query);
    const filteredPlugins = query.search ? filterPlugins(plugins, query.search) : plugins;

    return (
      <div className="page page-limited" id="marketplace-page">
        <Helmet title={translate('marketplace.page')} />
        <div className="page-notifs">
          {standaloneMode && (
            <PendingActions refreshPending={this.fetchPendingPlugins} pending={pending} />
          )}
        </div>
        <Header />
        <EditionBoxes
          canInstall={standaloneMode && !this.props.editionsReadOnly}
          canUninstall={standaloneMode}
          editions={editions}
          loading={this.props.loadingEditions}
          editionStatus={editionStatus}
          updateCenterActive={this.props.updateCenterActive}
          updateEditionStatus={this.props.setEditionStatus}
        />
        <Search
          query={query}
          updateCenterActive={this.props.updateCenterActive}
          updateQuery={this.updateQuery}
        />
        {loadingPlugins && <i className="spinner" />}
        {!loadingPlugins && (
          <PluginsList
            plugins={filteredPlugins}
            pending={pending}
            readOnly={!standaloneMode}
            refreshPending={this.fetchPendingPlugins}
            updateQuery={this.updateQuery}
          />
        )}
        {!loadingPlugins && <Footer total={filteredPlugins.length} />}
      </div>
    );
  }
}
