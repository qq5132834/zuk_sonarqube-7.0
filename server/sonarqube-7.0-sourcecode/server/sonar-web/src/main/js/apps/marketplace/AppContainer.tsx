/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { connect } from 'react-redux';
import App from './App';
import {
  getAppState,
  getGlobalSettingValue,
  getMarketplaceState,
  getMarketplaceEditions,
  getMarketplaceEditionStatus
} from '../../store/rootReducer';
import { Edition, EditionStatus } from '../../api/marketplace';
import { setEditionStatus } from '../../store/marketplace/actions';
import { RawQuery } from '../../helpers/query';

interface OwnProps {
  location: { pathname: string; query: RawQuery };
}

interface StateToProps {
  editions?: Edition[];
  editionsReadOnly: boolean;
  editionStatus?: EditionStatus;
  loadingEditions: boolean;
  standaloneMode: boolean;
  updateCenterActive: boolean;
}

interface DispatchToProps {
  setEditionStatus: (editionStatus: EditionStatus) => void;
}

const mapStateToProps = (state: any) => ({
  editions: getMarketplaceEditions(state),
  editionsReadOnly: getMarketplaceState(state).readOnly,
  editionStatus: getMarketplaceEditionStatus(state),
  loadingEditions: getMarketplaceState(state).loading,
  standaloneMode: getAppState(state).standalone,
  updateCenterActive:
    (getGlobalSettingValue(state, 'sonar.updatecenter.activate') || {}).value === 'true'
});

const mapDispatchToProps = { setEditionStatus };

export default connect<StateToProps, DispatchToProps, OwnProps>(
  mapStateToProps,
  mapDispatchToProps
)(App);
