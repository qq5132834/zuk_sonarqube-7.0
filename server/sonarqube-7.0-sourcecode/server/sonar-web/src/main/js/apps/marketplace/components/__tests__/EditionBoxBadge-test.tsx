/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import { EditionStatus } from '../../../../api/marketplace';
import EditionBoxBadge from '../EditionBoxBadge';

const DEFAULT_STATUS: EditionStatus = {
  currentEditionKey: '',
  nextEditionKey: '',
  installationStatus: 'NONE'
};

it('should display installed badge', () => {
  expect(
    getWrapper({
      status: {
        currentEditionKey: 'foo',
        nextEditionKey: '',
        installationStatus: 'NONE'
      }
    })
  ).toMatchSnapshot();
});

it('should display installing badge', () => {
  expect(
    getWrapper({
      status: {
        currentEditionKey: 'foo',
        nextEditionKey: 'foo',
        installationStatus: 'AUTOMATIC_IN_PROGRESS'
      }
    })
  ).toMatchSnapshot();
});

it('should display pending badge', () => {
  expect(
    getWrapper({
      status: {
        currentEditionKey: '',
        nextEditionKey: 'foo',
        installationStatus: 'AUTOMATIC_READY'
      }
    })
  ).toMatchSnapshot();
});

it('should not display a badge', () => {
  expect(
    getWrapper({
      status: {
        currentEditionKey: '',
        nextEditionKey: 'bar',
        installationStatus: 'AUTOMATIC_READY'
      }
    }).type()
  ).toBeNull();
});

function getWrapper(props = {}) {
  return shallow(<EditionBoxBadge editionKey="foo" status={DEFAULT_STATUS} {...props} />);
}
