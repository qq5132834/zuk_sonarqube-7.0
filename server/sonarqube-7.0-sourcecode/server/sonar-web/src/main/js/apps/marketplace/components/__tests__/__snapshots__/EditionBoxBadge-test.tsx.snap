// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display installed badge 1`] = `
<span
  className="marketplace-edition-badge badge badge-normal-size display-flex-center"
>
  <CheckIcon
    className="little-spacer-right"
    size={14}
  />
  marketplace.installed
</span>
`;

exports[`should display installing badge 1`] = `
<span
  className="marketplace-edition-badge badge badge-normal-size"
>
  marketplace.installing
</span>
`;

exports[`should display pending badge 1`] = `
<span
  className="marketplace-edition-badge badge badge-normal-size"
>
  marketplace.pending
</span>
`;
